gobblin-core/src/main/java/gobblin/converter/filter/AvroSchemaFieldRemover.java

```diff
@@ -0,0 +1,133 @@
+/* (c) 2015 LinkedIn Corp. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use
+ * this file except in compliance with the License. You may obtain a copy of the
+ * License at  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed
+ * under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+ * CONDITIONS OF ANY KIND, either express or implied.
+ */
+
+package gobblin.converter.filter;
+
+import java.util.List;
+import java.util.Map;
+
+import org.apache.avro.Schema;
+import org.apache.avro.Schema.Field;
+
+import com.google.common.base.Splitter;
+import com.google.common.collect.Lists;
+import com.google.common.collect.Maps;
+
+
+/**
+ * A class that removes specific fields from a (possibly recursive) Avro schema.
+ *
+ * @author ziliu
+ */
+public class AvroSchemaFieldRemover {
+
+  private final Map<String, AvroSchemaFieldRemover> children;
+  private final Map<String, Schema> schemaMap;
+
+  /**
+   * @param fieldNames Field names to be removed from the Avro schema. Contains comma-separated fully-qualified 
+   * field names, e.g., "header.memberId,mobileHeader.osVersion".
+   */
+  public AvroSchemaFieldRemover(String fieldNames) {
+    this();
+    this.addChildren(fieldNames);
+  }
+
+  private AvroSchemaFieldRemover() {
+    this.children = Maps.newHashMap();
+    this.schemaMap = Maps.newHashMap();
+  }
+
+  private void addChildren(String fieldNames) {
+    for (String fieldName : Splitter.on(',').trimResults().omitEmptyStrings().splitToList(fieldNames)) {
+      List<String> fieldNameComponents = Splitter.on('.').trimResults().omitEmptyStrings().splitToList(fieldName);
+      if (!fieldNameComponents.isEmpty()) {
+        this.addChildren(fieldNameComponents, 0);
+      }
+    }
+  }
+
+  private void addChildren(List<String> fieldNameComponents, int level) {
+    if (!this.children.containsKey(fieldNameComponents.get(level))) {
```

Comment:
`containsKey` may throw an NPE if the key is null, which may be the case if `fieldNameComponents` does not contain `level`.